### DISCUSSION TITLE
Ability to add multiple authors to a post

### DISCUSSION THREAD

**Discussion by sublimemarch** [forem/forem#15324] [Feature Requests]
As someone who frequently collaborates with other devs to write content, I'd like the ability to add other authors to a post. Currently, one of us submits an article and mentions the others as authors someone in the text of the post, but this doesn't make it apparent that the authors share equal credit.

This feature would likely include:
- The ability to add authors during the post's creation or after it has been published by username.
- An approval mechanism for authors, so that you can't be added as an author of a post without your consent.
- Posts showing on a co-author's profile just as if they were the "primary" author who submitted the post.
- Post authors would all show with their profile pic and name on an article, up to a certain amount (for example, 5 authors). Beyond a certain amount, text such as "and 3 other authors" would appear, and a user would need to click on this to see the rest. This is to maintain current styling and for ease of readability.

I don't think it would be necessary to change the "primary author" on a post (aka the person who originally submitted it) at this point in time, as this would likely require a different database schema.

**Comment by jessleenyc** [comment]
Thanks for the request @sublimemarch  -- we can actually add two additional authors onto a post as an admin. We don't have all the features you detailed yet, but are there specific posts you'd like us to update for now?

**Reply by trusktr** [reply]
Please add this!

**Reply by Lisa339** [reply]
thank you for sharing this information.

**Comment by benhalpern** [comment]
I could have sworn there was an issue for this since we've definitely talked about the implementation of this at some point, am I imagining that?

Either way, definitely agree.

**Comment by Zhao-Andy** [comment]
...I meant to make an issue about that but I didn't, oops. 😅

Totally agree with all your points @sublimemarch. In terms of design, there are a few places we'll need to consider to show the additional author(s):
- Top of the page
- Author sidebar
- Bottom of the page

Here are some screenshots of what that might look like:

![two authors](https://user-images.githubusercontent.com/17884966/49771094-8aa2b800-fcb5-11e8-8710-7bdc706d21ec.png)
Bottom of page:

![bottom of page](https://user-images.githubusercontent.com/17884966/49808256-21a95780-fd2a-11e8-873f-1261905684be.png)

**Comment by Nerajno** [comment]
Has it been addressed ?

**Comment by rhymes** [comment]
Hi @Nerajno! No, not really :) We don't have an immediate plan to address this but it is open for third party contributors.

I suggest anyone who wants to tackle this to kickstart a conversation on how they'd like to implement it (in the code and visually) before diving solo in creating a PR!

**Comment by jimmoffitt** [comment]
I am starting to stage a new post, and wondering how to get my colleague added as an author. Any tips? Do you just need a github handle?

**Comment by vimota** [comment]
Hi, I'm collaborating on a teammate on a post to Dev.to - is an admin able to add me to the post as a contributor? Thank you!

**Comment by krishnadevz** [comment]
In Dev.to now hackathon is started so some folks are working alone or some are working in teams so for that dev can have add contributors thing like GitHub is having so both members can add and contribute to articles according to their own work in project/hackathon. and hello ben and other developers please think about  I think this is helpful for many folks.
Like this 🔽

![Annotation 2020-05-18 102213](https://user-images.githubusercontent.com/42638797/82175413-783c9d00-98f1-11ea-8909-671d6722927e.png)

**Comment by yozlet** [comment]
(Caveat: I'm an interested beginner to the codebase, but I have Rails experience)

Looking at @sublimemarch's suggested design, I feel like the hardest part is the approval/consent flow. Here's an attempt at a design that might not be too hard:

1. Primary author drafts article, adds a `coauthor: USER1 USER2` frontmatter line and saves it
2. Primary author shares the article URL with co-authors
3. Co-authors go to URL, see "Add yourself as co-author?" button, click it. This adds a record to a `coauthor_approval` join table, or something similar
4. Article view adds co-authors who completed approval

### Abuse potential

1. Author invites co-author to approve article about conversation between the two of them.
2. Co-author approves.
3. Author updates article to be negative/abusive.

(Does the existing admin-driven addition of `second_user_id` & `third_user_id` have this potential too?)

So, the article view should check whether the article has changed since the co-author approved. This may require more than just checking the date of last update, since the update might just have been setting `published` to `true`.

As I said, I'm interested in helping with this, but given the constraints of my day job I may not have much time. I'm primarily throwing this in to help identify work to be done.

**Comment by tylerauerbeck** [comment]
Has any work been started on this? I'm interested in seeing this feature added and (sadly) noticed that the future label was removed. I'd be happy to start hacking around on this and would be happy to pick up on any exising work if there is any.

**Comment by yozlet** [comment]
@tylerauerbeck Given that it's pretty quiet on this thread, maybe the best way is to try starting a PR? It doesn't need to be complete, just enough to get more solid feedback. I haven't done any Rails in a couple of years, but I can offer a couple of hours to make tests and that kind of thing.

CCing @rhymes in case they have more to say on the `future` label removal.

**Comment by rhymes** [comment]
@yozlet we removed the `future` label entirely because it was a bit unclear and we'de prefer to rely on approved features vs non approved. I think coming up with a plan or MVP PR is probably easier yes. Adding the concept of co-authors is a bit tricky for all the reasons you've enumeated in the past @yozlet

**Comment by tylerauerbeck** [comment]
@yozlet @rhymes Thanks for the details. This was mostly trying to feel out where things currently stand and see if I was missing a larger reason that this wasn’t being worked. I’m happy to begin an MVP PR here as I have time.

**Comment by michael-tharrington** [comment]
Another interesting thing to consider here. When we post an article by multiple authors, the article should be promoted via [/notifications](https://dev.to/notifications) to the followers of each author. 

That said, if a user happens to be following multiple authors of a single post, the post should only be promoted to this user once.

**Comment by mstruve** [comment]
@juliannatetreault is going to champion an RFC for this 🙌

**Reply by KartikSoneji** [reply]
Hi @juliannatetreault, is there somewhere I can see and comment on the RFC?

**Comment by askpt** [comment]
This is actually an important feature since I collaborate with other people and they also should receive their credit.

**Comment by GhaythFuad** [comment]
+1, this is needed

**Comment by hnrq** [comment]
+1!! This would be great to have!

**Comment by anishamalde** [comment]
+ 1 we would love this!
